proxy/proxy.go

```diff
@@ -280,13 +280,17 @@ func (proxy *Proxy) Listen() []net.Listener {
 	return listeners
 }
 
-func (proxy *Proxy) Serve(listeners []net.Listener) {
+func (proxy *Proxy) Serve(listeners []net.Listener, ready func()) {
 	errs := make(chan error)
 	for _, listener := range listeners {
 		go func(listener net.Listener) {
 			errs <- (&http.Server{Handler: proxy}).Serve(listener)
 		}(listener)
 	}
+	// It would be better if we could delay calling Done() until all
+	// the listeners are ready, but it doesn't seem to be possible to
+	// hook the right point in http.Server
```

Comment:
I looked at Go's sources and also couldn't find anything. This is indeed a bit annoying.
Could something like the below help to reduce the window for a race, even if still not strictly correct?

```
func (proxy *Proxy) Serve(listeners []net.Listener) {
	created := make(chan bool)
	errs := make(chan error)
	for _, listener := range listeners {
		go func(listener net.Listener) {
			server := http.Server{Handler: proxy}
			created <- true
			errs <- server.Serve(listener)
		}(listener)
	}
	for range listeners {
		<-created
	}
	ready()

	for range listeners {
		err := <-errs
		if err != nil {
			Log.Fatalf("Serve failed: %s", err)
		}
	}
}
```